src/boot_O2_g3/yaz0.c

```diff
@@ -0,0 +1,150 @@
+#include <ultra64.h>
+#include <global.h>
+
+u8 sYaz0DataBuffer[0x400];
+u8* sYaz0CurDataEnd;
+u32 sYaz0CurRomStart;
+u32 sYaz0CurSize;
+u8* sYaz0MaxPtr;
+u8* D_8009BE20;
+
+void* Yaz0_FirstDMA() {
+    u32 pad0;
+    u32 dmaSize;
+    u32 curSize;
+
+    sYaz0MaxPtr = sYaz0CurDataEnd - 0x19;
+
+    curSize = (u32)sYaz0CurDataEnd - (u32)sYaz0DataBuffer;
+    dmaSize = (curSize > sYaz0CurSize) ? sYaz0CurSize : curSize;
+
+    DmaMgr_DMARomToRam(sYaz0CurRomStart, sYaz0DataBuffer, dmaSize);
+    sYaz0CurRomStart += dmaSize;
+    sYaz0CurSize -= dmaSize;
+    return sYaz0DataBuffer;
+}
+
+void* Yaz0_NextDMA(void* curSrcPos) {
+    u8* dst;
+    u32 restSize;
+    u32 dmaSize;
+    OSPri oldPri;
+
+    restSize = (u32)sYaz0CurDataEnd - (u32)curSrcPos;
+
+    dst = (restSize & 7) ? (sYaz0DataBuffer - (restSize & 7)) + 8 : sYaz0DataBuffer;
+
+    bcopy(curSrcPos, dst, restSize);
+    dmaSize = ((u32)sYaz0CurDataEnd - (u32)dst) - restSize;
+    if (sYaz0CurSize < dmaSize) {
+        dmaSize = sYaz0CurSize;
+    }
+
+    if (dmaSize != 0) {
+        DmaMgr_DMARomToRam(sYaz0CurRomStart, dst + restSize, dmaSize);
+        sYaz0CurRomStart += dmaSize;
+        sYaz0CurSize -= dmaSize;
+        if (!sYaz0CurSize) {
+            sYaz0MaxPtr = dst + restSize + dmaSize;
+        }
+    } else {
+        oldPri = osGetThreadPri(NULL);
+        osSetThreadPri(NULL, 0x7F);
+        Fault_Log("圧縮展開異常\n");
+        osSetThreadPri(NULL, oldPri);
+    }
+
+    return dst;
+}
+
+typedef struct {
+    /* 0x00 */ u32 magic; // Yaz0
+    /* 0x04 */ u32 decSize;
+    /* 0x08 */ u32 compInfoOffset; // only used in mio0
+    /* 0x0C */ u32 uncompDataOffset; // only used in mio0
+} Yaz0Header; // size = 0x10
+
+#define YAZ0_MAGIC 0x59617A30 // "Yaz0"
+
+s32 Yaz0_DecompressImpl(u8* src, u8* dst) {
+    u32 bitIdx = 0;
+    u8* dstEnd;
+    u32 chunkHeader = 0;
+    u32 nibble;
+    u8* backPtr;
+    s32 chunkSize;
+    u32 off;
+    u32 magic;
+
+    magic = ((Yaz0Header*)src)->magic;
```

Comment:
Is this required for matching, or maybe warning suppression? OoT lets `Yaz0_DecompressImpl` take a `Yaz0Header*` instead of `u8*`. Why might we be adverse to doing the same?

Reply:
Required for matching. I couldn't get it to match without reusing the `src` argument, which happens a couple of lines down.

Reply:
Ahh I see it now, got it.